File: cms/src/main/java/com/newtv/cms/contract/SearchContract.java
package com.newtv.cms.contract;

import android.content.Context;

import com.newtv.cms.CmsServicePresenter;
import com.newtv.cms.DataObserver;
import com.newtv.cms.ICmsPresenter;
import com.newtv.cms.ICmsView;
import com.newtv.cms.api.ISearch;
import com.newtv.cms.bean.ModelResult;
import com.newtv.cms.bean.SubContent;
import com.newtv.libs.Libs;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * 项目名称:         CBoxTV2.0
 * 包名:            tv.newtv.contract
 * 创建事件:         14:44
 * 创建人:           weihaichao
 * 创建日期:          2018/10/16
 */
public class SearchContract {
    public interface View extends ICmsView {
        void searchResult(List<SubContent> result);
    }

    public interface Presenter extends ICmsPresenter {
        /**
         * 搜索
         */
        void search(SearchCondition condition);
    }

    public static class SearchCondition {

        String categoryId = "";
        String contentType = "";
        String videoType = "";
        String videoClass = "";
        String area = "";
        String year = "";
        String keyword = "";
        String page = "0";          //第几页
        String rows = "40";         //每页条数
        String keywordType = "";

        /**
         * @return
         */
        public static SearchCondition Builder() {
            return new SearchCondition();
        }

        public SearchCondition setRows(String rows) {
            this.rows = rows;
            return this;
        }

        public SearchCondition setPage(String page) {
            this.page = page;
            return this;
        }

        public SearchCondition setKeywordType(String keywordType) {
            this.keywordType = keywordType;
            return this;
        }

        public SearchCondition setKeyword(String keyword) {
            this.keyword = keyword;
            return this;
        }

        public SearchCondition setYear(String year) {
            this.year = year;
            return this;
        }

        public SearchCondition setVideoType(String videoType) {
            this.videoType = videoType;
            return this;
        }

        public SearchCondition setVideoClass(String videoClass) {
            this.videoClass = videoClass;
            return this;
        }

        public SearchCondition setContentType(String contentType) {
            this.contentType = contentType;
            return this;
        }

        public SearchCondition setCategoryId(String categoryId) {
            this.categoryId = categoryId;
            return this;
        }

        public SearchCondition setArea(String area) {
            this.area = area;
            return this;
        }
    }

    public static class SearchPresenter extends CmsServicePresenter<View> implements Presenter {

        public SearchPresenter(@NotNull Context context, @NotNull View view) {
            super(context, view);
        }

        @Override
        public void search(SearchCondition condition) {
            ISearch search = getService(SERVICE_SEARCH);
            if (search != null) {
                search.search(
                        Libs.get().getAppKey(),
                        Libs.get().getChannelId(), condition.categoryId,
                        condition.contentType, condition.videoType, condition.videoClass, condition
                                .area, condition.year, condition.keyword, condition.page, condition
                                .rows, condition.keywordType, new
                                DataObserver<ModelResult<List<SubContent>>>() {
                                    @Override
                                    public void onResult(ModelResult<List<SubContent>> result) {
                                        if (result.isOk()) {
                                            getView().searchResult(result.getData());
                                        } else {
                                            getView().onError(getContext(), result.getErrorMessage());
                                        }
                                    }

                                    @Override
                                    public void onError(@Nullable String desc) {
                                        getView().onError(getContext(), desc);
                                    }
                                });
            }
        }
    }
}

File: cms/src/main/java/com/newtv/cms/contract/SearchContract.kt
package com.newtv.cms.contract

import android.content.Context
import com.newtv.cms.CmsServicePresenter
import com.newtv.cms.DataObserver
import com.newtv.cms.ICmsPresenter
import com.newtv.cms.ICmsView
import com.newtv.cms.api.ISearch
import com.newtv.cms.bean.ModelResult
import com.newtv.cms.bean.SubContent
import com.newtv.libs.Libs
import java.util.*


/**
 * 项目名称:         CBoxTV2.0
 * 包名:            tv.newtv.contract
 * 创建事件:         14:44
 * 创建人:           weihaichao
 * 创建日期:          2018/10/16
 */
class SearchContract {
    interface View : ICmsView {
        fun searchResult(requestID: Long, result: ArrayList<SubContent>?)
    }

    interface LoadingView : View {
        fun onLoading()
        fun loadingFinish()
    }

    interface Presenter : ICmsPresenter {
        /**
         * 搜索
         */
        fun search(condition: SearchCondition): Long
    }

    class SearchCondition {

        internal var categoryId = ""
        internal var contentType = ""
        internal var videoType = ""
        internal var videoClass = ""
        internal var area = ""
        internal var year = ""
        internal var keyword = ""
        internal var page = "0"          //第几页
        internal var rows = "40"         //每页条数
        internal var keywordType = ""

        fun setRows(rows: String): SearchCondition {
            this.rows = rows
            return this
        }

        fun setPage(page: String): SearchCondition {
            this.page = page
            return this
        }

        fun setKeywordType(keywordType: String): SearchCondition {
            this.keywordType = keywordType
            return this
        }

        fun setKeyword(keyword: String): SearchCondition {
            this.keyword = keyword
            return this
        }

        fun setYear(year: String): SearchCondition {
            this.year = year
            return this
        }

        fun setVideoType(videoType: String): SearchCondition {
            this.videoType = videoType
            return this
        }

        fun setVideoClass(videoClass: String): SearchCondition {
            this.videoClass = videoClass
            return this
        }

        fun setContentType(contentType: String): SearchCondition {
            this.contentType = contentType
            return this
        }

        fun setCategoryId(categoryId: String): SearchCondition {
            this.categoryId = categoryId
            return this
        }

        fun setArea(area: String): SearchCondition {
            this.area = area
            return this
        }

        companion object {
            /**
             * @return
             */
            @JvmStatic
            fun Builder(): SearchCondition {
                return SearchCondition()
            }
        }
    }

    class SearchPresenter(context: Context, view: View)
        : CmsServicePresenter<View>(context, view), Presenter {

        override fun search(condition: SearchCondition): Long {
            val search = getService<ISearch>(CmsServicePresenter.SERVICE_SEARCH)
            search?.let { iSearch ->
                view?.let {
                    if (it is LoadingView) {
                        it.onLoading()
                    }
                }
                val requestID: Long = iSearch.search(
                        Libs.get().appKey,
                        Libs.get().channelId, condition.categoryId,
                        condition.contentType, condition.videoType, condition.videoClass, condition
                        .area, condition.year, condition.keyword, condition.page, condition
                        .rows, condition.keywordType, object : DataObserver<ModelResult<ArrayList<SubContent>>> {
                    override fun onResult(result: ModelResult<ArrayList<SubContent>>, requestCode: Long) {
                        if (result.isOk()) {
                            view?.let {
                                it.searchResult(requestCode, result.data)
                                if (it is LoadingView) {
                                    it.loadingFinish()
                                }
                            }
                        } else {
                            view?.onError(context, result.errorMessage)
                        }
                    }

                    override fun onError(desc: String?) {
                        view?.let { callback ->
                            callback.onError(context, desc)
                            if (callback is LoadingView) {
                                callback.loadingFinish()
                            }
                        }
                    }
                })
                return requestID
            }
            return 0
        }
    }
}
